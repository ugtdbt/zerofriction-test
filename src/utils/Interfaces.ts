export interface State<T> {
  data: T;
  valideAddressForm: boolean;
  valideContactDetailsForm: boolean;
  valideOrganizationDetailsForm: boolean;
  success: boolean;
  save: boolean;
  cancel: boolean;
  formtouched: boolean;
  formreset: boolean;
}

export interface Action<T> {
  type: string;
  payload?: T;
}
