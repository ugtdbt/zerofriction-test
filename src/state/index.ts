export * as actionCreators from "./actionCreators/organisationConfigurationActionCreator";
export * from "./store";
export * from "./reducers/index";
