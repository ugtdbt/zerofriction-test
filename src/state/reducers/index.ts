import { combineReducers } from "redux";
import organisationConfigurationReducer from "./organisationConfigurationReducer";

const reducers = combineReducers({
  organisationConfiguration: organisationConfigurationReducer,
});

export default reducers;

export type State = ReturnType<typeof reducers>;
