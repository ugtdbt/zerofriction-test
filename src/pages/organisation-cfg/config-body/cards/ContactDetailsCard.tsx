import React, { memo, useState, useEffect } from "react";
import { Card, Form, Input } from "antd";
import type { State } from "../../../../state";
import { useSelector, useDispatch } from "react-redux";
import type { ContactDetailType } from "../../../../utils/types/organisation-config";
import type { FormInstance } from "antd/es/form";
import { bindActionCreators } from "redux";
import { actionCreators } from "../../../../state";

const ContactDetailsCard: React.FC = memo(() => {
  const formRef = React.useRef<FormInstance>(null);
  const emailRegex = /\S+@\S+\.\S+/;
  const telephoneRegex = /^[0-9\+]{1,11}$/;

  const initialValidationData = {
    emailAddress: {
      validateStatus: "",
      help: "",
    },
    telephone: {
      validateStatus: "",
      help: "",
    },
    website: {
      validateStatus: "",
      help: "",
    },
  };

  const { data, formreset, save } = useSelector(
    (state: State) => state.organisationConfiguration
  );

  const dispatch = useDispatch();
  const { valideContactDetails, clickContactDetailsSaveButton } =
    bindActionCreators(actionCreators, dispatch);

  useEffect(() => {
    formRef.current?.resetFields();
    setFormErrors(initialValidationData);
    setFormDisabled(false);
  }, [formreset]);

  useEffect(() => {
    if (save) {
      clickContactDetailsSaveButton(formValues);
      setFormDisabled(true);
    }
  }, [save]);

  const [formValues, setFormValues] = useState<ContactDetailType>(
    data.organizationConfig.contactDetails
  );
  const [formErrors, setFormErrors] = useState(initialValidationData);
  const [formDisabled, setFormDisabled] = useState<boolean>(false);

  const handleContactDetailsChnage = (e: any) => {
    const { id, value } = e.target;
    setFormValues({ ...formValues, [id]: value });
    validate({ ...formValues, [id]: value }, id);
    validateForm({ ...formValues, [id]: value });
  };

  const validateForm = (formValues: ContactDetailType) => {
    let fromValide = true;
    if (formValues["website"] === "") {
      fromValide = false;
    }

    if (formValues["emailAddress"] === "") {
      fromValide = false;
    } else if (!emailRegex.test(formValues["emailAddress"])) {
      fromValide = false;
    }

    if (formValues["telephone"] === "") {
      fromValide = false;
    } else if (!telephoneRegex.test(formValues["telephone"])) {
      fromValide = false;
    }

    valideContactDetails(fromValide);
  };

  const validate = (formValues: ContactDetailType, key: string) => {
    switch (key) {
      case "website":
        if (formValues[key] === "") {
          setFormErrors({
            ...formErrors,
            website: {
              validateStatus: "error",
              help: "website is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            website: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "emailAddress":
        if (formValues[key] === "") {
          setFormErrors({
            ...formErrors,
            emailAddress: {
              validateStatus: "error",
              help: "emailAddress is required!",
            },
          });
        } else if (!emailRegex.test(formValues[key])) {
          setFormErrors({
            ...formErrors,
            emailAddress: {
              validateStatus: "error",
              help: "Email address is not valid!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            emailAddress: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "telephone":
        if (formValues[key] === "") {
          setFormErrors({
            ...formErrors,
            telephone: {
              validateStatus: "error",
              help: "telephone is required!",
            },
          });
        } else if (!telephoneRegex.test(formValues[key])) {
          setFormErrors({
            ...formErrors,
            telephone: {
              validateStatus: "error",
              help: "Telephone number is not valid!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            telephone: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      default:
        break;
    }
  };

  return (
    <Card title="Contact Details">
      <Form
        labelCol={{ span: 10 }}
        wrapperCol={{ span: 14 }}
        layout="horizontal"
        style={{ maxWidth: 600 }}
        ref={formRef}
        disabled={formDisabled}
      >
        <Form.Item
          label="EmailAddress"
          name="emailAddress"
          validateStatus={
            formErrors.emailAddress.validateStatus === "error"
              ? "error"
              : undefined
          }
          help={
            formErrors.emailAddress.help !== ""
              ? formErrors.emailAddress.help
              : undefined
          }
        >
          <Input onChange={handleContactDetailsChnage} />
        </Form.Item>
        <Form.Item
          label="Telephone"
          name="telephone"
          validateStatus={
            formErrors.telephone.validateStatus === "error"
              ? "error"
              : undefined
          }
          help={
            formErrors.telephone.help !== ""
              ? formErrors.telephone.help
              : undefined
          }
        >
          <Input style={{ width: 260 }} onChange={handleContactDetailsChnage} />
        </Form.Item>
        <Form.Item
          label="website"
          name="website"
          validateStatus={
            formErrors.website.validateStatus === "error" ? "error" : undefined
          }
          help={
            formErrors.website.help !== "" ? formErrors.website.help : undefined
          }
        >
          <Input onChange={handleContactDetailsChnage} />
        </Form.Item>
      </Form>
    </Card>
  );
});

ContactDetailsCard.displayName = "ContactDetailsCard";
export default ContactDetailsCard;
