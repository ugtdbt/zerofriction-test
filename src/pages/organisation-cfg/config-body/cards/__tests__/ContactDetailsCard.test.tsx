import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import ContactDetailsCard from "../ContactDetailsCard";
import { store } from "../../../../../state/store";
import { Provider } from "react-redux";
import "../../../../../utils/matchMedia";

describe("ContactDetailsCard component", () => {
  const ContactDetailsCardComponent = (
    <React.StrictMode>
      <Provider store={store}>
        <ContactDetailsCard />
      </Provider>
    </React.StrictMode>
  );

  afterEach(() => {
    cleanup();
  });

  it("ContactDetailsCardComponent Match Snapshot", () => {
    const tree = renderer.create(ContactDetailsCardComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should render ContactDetailsCardComponent component", () => {
    render(ContactDetailsCardComponent);
    const mainElem = screen.getByTestId("contact-details-form");
    expect(mainElem).toBeInTheDocument();
  });
});
