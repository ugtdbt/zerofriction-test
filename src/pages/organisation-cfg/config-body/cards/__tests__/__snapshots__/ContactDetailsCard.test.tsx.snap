// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactDetailsCard component ContactDetailsCardComponent Match Snapshot 1`] = `
<div
  className="ant-card ant-card-bordered css-dev-only-do-not-override-1km3mtt"
  data-testid="contact-details-form"
>
  <div
    className="ant-card-head"
    style={Object {}}
  >
    <div
      className="ant-card-head-wrapper"
    >
      <div
        className="ant-card-head-title"
      >
        Contact Details
      </div>
    </div>
  </div>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <form
      className="ant-form ant-form-horizontal css-dev-only-do-not-override-1km3mtt"
      onReset={[Function]}
      onSubmit={[Function]}
      style={
        Object {
          "maxWidth": 600,
        }
      }
    >
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="emailAddress"
              title="EmailAddress"
            >
              EmailAddress
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="emailAddress"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="telephone"
              title="Telephone"
            >
              Telephone
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="telephone"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={
                    Object {
                      "width": "100%",
                    }
                  }
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="website"
              title="website"
            >
              website
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="website"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
